File: ViuaBasic/Variables.cs
using System;
using System.Collections.Generic;

namespace ViuaBasic
{
  public class Variables
  {
    public enum Type
    {
      UNKNOWN,
      INTEGER,
      FLOAT,
      STRING,
      ARRAY
    }

    private Dictionary<string, int> names;
    private Dictionary<string, Type> types;
    private Dictionary<string, Type> array_types;

    public Variables()
    {
      names = new Dictionary<string, int>();
      types = new Dictionary<string, Type>();
      array_types = new Dictionary<string, Type>();
    }

    public bool exists(string name)
    {
      return names.ContainsKey(name);
    }

    public bool is_array(string name)
    {
      bool result = exists(name);
      if (result)
      {
        result = get_type(name).Equals(Type.ARRAY);
      }
      return result;
    }

    public void set_var(string name, Type type, int register)
    {
      if (exists(name))
      {
        throw new Exception("var " + name + " exists");
      }
      names.Add(name, register);
      types.Add(name, type);
    }

    public void set_array(string name, Type type, int register)
    {
      if (exists(name))
      {
        throw new Exception("var " + name + " exists");
      }
      names.Add(name, register);
      types.Add(name, Type.ARRAY);
      array_types.Add(name, type);
    }

    public int get_register(string name)
    {
      if (!exists(name))
      {
        throw new Exception("var " + name + " not exists");
      }
      return names[name];
    }

    public Type get_type(string name)
    {
      if (!exists(name))
      {
        throw new Exception("var " + name + " not exists");
      }
      return types[name];
    }

    public Type get_array_type(string name)
    {
      if (!exists(name))
      {
        throw new Exception("var " + name + " not exists");
      }
      if (!is_array(name))
      {
        throw new Exception("var " + name + " is not an array");
      }
      return array_types[name];
    }
  }
}

File: src/main/java/org/viuavm/viuabasic/Variables.java
package org.viuavm.viuabasic;

import java.util.HashMap;

public class Variables
{
  public enum Type
  {
    UNKNOWN,
    INTEGER,
    FLOAT,
    STRING,
    ARRAY
  }

  private HashMap<String, Integer> names;
  private HashMap<String, Type> types;
  private HashMap<String, Type> array_types;

  public Variables()
  {
    names = new HashMap<String, Integer>();
    types = new HashMap<String, Type>();
    array_types = new HashMap<String, Type>();
  }

  public boolean exists(String name)
  {
    return names.containsKey(name);
  }

  public boolean is_array(String name)
  {
    boolean result = exists(name);
    if (result)
    {
      result = get_type(name).equals(Type.ARRAY);
    }
    return result;
  }

  public void set_var(String name, Type type, int register)
  {
    if (exists(name))
    {
      throw new RuntimeException("var " + name + " exists");
    }
    names.put(name, register);
    types.put(name, type);
  }

  public void set_array(String name, Type type, int register)
  {
    if (exists(name))
    {
      throw new RuntimeException("var " + name + " exists");
    }
    names.put(name, register);
    types.put(name, Type.ARRAY);
    array_types.put(name, type);
  }

  public int get_register(String name)
  {
    if (!exists(name))
    {
      throw new RuntimeException("var " + name + " not exists");
    }
    return names.get(name);
  }

  public Type get_type(String name)
  {
    if (!exists(name))
    {
      throw new RuntimeException("var " + name + " not exists");
    }
    return types.get(name);
  }

  public Type get_array_type(String name)
  {
    if (!exists(name))
    {
      throw new RuntimeException("var " + name + " not exists");
    }
    if (!is_array(name))
    {
      throw new RuntimeException("var " + name + " is not an array");
    }
    return array_types.get(name);
  }
}
